File: RayVE.CSharp/Output/PPMDocument.cs
using System;
using System.Collections.Generic;

namespace RayVE.Output
{
    public class PPMDocument
    {
        private readonly Canvas _canvas;
        private readonly int _maxValue;

        private IEnumerable<string> GetPPMHeader(int maxValue)
            => new List<string>()
            {
                "P3",
                $"{_canvas.Width} {_canvas.Height}",
                $"{maxValue}"
            };

        public PPMDocument(Canvas canvas, int maxValue)
        {
            _canvas = canvas ?? new Canvas(0, 0);
            _maxValue = maxValue;
        }

        public override string ToString()
        {
            var rows = new List<string>();
            rows.AddRange(GetPPMHeader(_maxValue));

            for (var i = 0; i < _canvas.Height; i++)
                rows.AddRange(GetRowPPM(i, _maxValue));

            return String.Join(Environment.NewLine, rows) + Environment.NewLine;
        }

        private IEnumerable<string> GetRowPPM(int row, int maxValue)
        {
            var rowValues = new List<string>();

            var pixels = _canvas.GetPixels();

            for (var i = 0; i < _canvas.Width; i++)
                rowValues.AddRange(new PPMColor(pixels[i, row]).ToPPM(maxValue));

            var subRowValues = new List<string>();
            var subRowLength = 0;

            foreach (var value in rowValues)
            {
                if (subRowLength + value.Length + subRowValues.Count > 70)
                {
                    yield return String.Join(" ", subRowValues);
                    subRowValues = new List<string>();
                    subRowLength = 0;
                }

                subRowValues.Add(value);
                subRowLength += value.Length;
            }

            if (subRowValues.Count > 0)
                yield return String.Join(" ", subRowValues);
        }

    }
}

File: RayVE.FSharp/Output/PPMDocument.fs
namespace RayVE

open System

type PPMDocument = {
    Canvas: Canvas
    MaxValue: int
}

module PPMDocument =
    let header doc =
        [|
            "P3"
            $"{doc.Canvas.Width} {doc.Canvas.Height}"
            $"{doc.MaxValue}"
        |]

    let toPPM doc =
        let maxChunkSize = 70
        let width = doc.Canvas.Width
        let height = doc.Canvas.Height
        let pixels = doc.Canvas.Pixels
        let data = Array2D.init width height (fun i j -> Color.toPPM pixels.[i, j] doc.MaxValue)
                   |> Array2D.toJagged
                   |> Array.transpose
                   |> Array.map Array.concat
                   |> Array.map (String.chunkSplit maxChunkSize " ")
                   |> Seq.collect id
                   |> String.concat Environment.NewLine

        [| yield! header doc; yield data |]
        |> String.concat Environment.NewLine